platform-sdk/docs/proposals/TSS-Ledger-Id/TSS-Ledger-Id.md

```diff
@@ -0,0 +1,1027 @@
+for the node, the public shares, and the ledger id.
+
+Internal Elements:
+
+1. Map<RosterHash, Roster> rosters
+2. Map<RosterHash, List<TssMessage>> tssMessages
+3. Map<RosterHash, Record(RosterHash, List<PairingPrivateKey>, List<PairingPublicKey>, LedgerId)> shares
+
+Inputs:
+
+1. Pair<RosterHash, Roster> (Wire)
+2. Pair<RosterHash, List<TssMessage>> (Wire)
+
+Outputs
+
+1. Record(RosterHash, List<PairingPrivateKey>, List<PairingPublicKey>, LedgerId) (Wire)
+
+#### Transaction Resubmitter
+
+The transaction resubmitter is extended to detect which the new `TssMessageTransaction` and `TssVoteTransaction`
+system transactions have not reached consensus because the event containing them has become stale. The Transaction
+Resubmitter will resubmit these transactions to the transaction pool for reprocessing.
+
+### Configuration
+
+The following are new configuration:
+
+1. `tss.maxSharesPerNode` - The maximum number of shares that can be assigned to a node.
+
+### Metrics
+
+The following metrics should be added to the platform:
+
+1. When a candidate roster is set.
+2. The number of TSS Messages collected for a candidate roster.
+3. The number of votes collected for a candidate roster.
+4. The time it takes to compute shares from the key material.
+
+### Performance
+
+The `TSS State Manager` execution on the transaction handling through routes TSS related system transactions to the
+relevant components and updates the TSS related data structures in the state. The routing should be a negligible
+impact provided the receiving components receive the messages and handles them on their own thread. The state
+updates should be simple crud operations, adds and deletes to data structures. All other computation on the
+transaction handling thread should be basic bookkeeping calculations and TSS signature verifications.
+
+---
+
+## Test Plan
+
+### Unit Tests
+
+Apart from the obvious unit testing of methods and classes, the following scenarios should be unit tested:
+
+- Failure Scenarios
+  - Bad TssMessages
+  - Failure to reach threshold number of votes
+- HAPI Tests
+  - A down node during pre-genesis needing to get the key material from post-genesis node.
+  - Down nodes during bootstrap and re-keying
+  - Multi-release migrations
+
+### Integration Tests
+
+Additional Integration Tests:
+
+- TSS bootstrap for New Networks
+- TSS bootstrap for Existing Networks
+  - multiple software upgrades to check transfer of the ledger id.
+- Network State Transplant
+
+### Performance Tests
+
+The following needs performance profiling:
+
+1. How long does it take to generate the key material for a new candidate roster as a function of roster size?
+2. How much space is being used in the state for all TSS related data structures?
+
+---
+
+## Implementation and Delivery Plan
+
+### Implementation
+
+1. The protobufs for the new system transactions and state data structures.
```

Comment:
Might be worth calling out that this includes modifications to the BlockStream protobufs.

There should be a HIP for this proposal too, and all protobuf changes needed should be included in that HIP.